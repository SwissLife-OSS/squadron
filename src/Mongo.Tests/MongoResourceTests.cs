using System.Collections.Generic;
using System.IO;
using System.Threading.Tasks;
using MongoDB.Bson;
using MongoDB.Driver;
using Snapshooter.Xunit;
using Xunit;

namespace Squadron
{
    public class MongoResourceTests
        : IClassFixture<MongoResource>
    {
        public MongoResourceTests(MongoResource mongoResource)
        {
            MongoResource = mongoResource;
        }

        private MongoResource MongoResource { get; }

        [Fact]
        public void CreateAndUseCollection()
        {
            // arrange
            IMongoCollection<BsonDocument> collection = MongoResource.CreateCollection<BsonDocument>();

            BsonDocument document = new BsonDocument {{"a", "b"}};

            // act
            collection.InsertOne(document);

            // assert
            BsonDocument retrieved = collection.Find(new BsonDocument()).First();
            Assert.True(retrieved.Contains("a"));
            Assert.Equal(document["a"], retrieved["a"]);
        }

        [Fact]
        public void ReUseDatabase()
        {
            // arrange
            BsonDocument document = new BsonDocument {{"a", "b"}};

            // act
            IMongoDatabase database = MongoResource.CreateDatabase();
            IMongoCollection<BsonDocument> collectiona = MongoResource.CreateCollection<BsonDocument>(database);
            IMongoCollection<BsonDocument> collectionb = MongoResource.CreateCollection<BsonDocument>(database);

            collectiona.InsertOne(document);
            collectionb.InsertOne(document);

            // assert
            Assert.NotEqual(collectiona, collectionb);
            Assert.Equal(database, collectiona.Database);
            Assert.Equal(database, collectionb.Database);

            BsonDocument retrieveda = collectiona.Find(new BsonDocument()).First();
            BsonDocument retrievedb = collectionb.Find(new BsonDocument()).First();

            Assert.True(retrieveda.Contains("a"));
            Assert.Equal(document["a"], retrieveda["a"]);
            Assert.True(retrievedb.Contains("a"));
            Assert.Equal(document["a"], retrievedb["a"]);
        }

        [Fact]
        public async Task CreateCollectionFromFile_DefaultOptions()
        {
            // act
            IMongoCollection<BsonDocument> collection = await MongoResource.CreateCollectionFromFileAsync<BsonDocument>(
                null,
                new CreateCollectionFromFileOptions
                {
                    File = new FileInfo(Path.Combine("Resources", "news.json")),
                });

            // assert
            BsonDocument imported = collection.Find(new BsonDocument()).FirstOrDefault();
            imported.MatchSnapshot(o => o.IgnoreField("[0].Value"));
        }

        [Fact]
        public async Task CreateDatabaseFromFiles()
        {
            // act
            IMongoDatabase db = await MongoResource.CreateDatabase(
                new FileInfo(Path.Combine("Resources", "news.json")));

            // assert
            IMongoCollection<BsonDocument> imported = db.GetCollection<BsonDocument>("news");
            List<BsonDocument> items = await imported
                .Find(Builders<BsonDocument>.Filter.Empty)
                .ToListAsync();
            items.MatchSnapshot(o => o.IgnoreField("[0].[0].Value"));
        }

        [Fact]
        public async Task CreateCollectionFromFile_CustomOptions()
        {
            // act
            IMongoCollection<BsonDocument> collection = await MongoResource.CreateCollectionFromFileAsync<BsonDocument>(
                null,
                new CreateCollectionFromFileOptions
                {
                    File = new FileInfo(Path.Combine("Resources", "news.tsv")),
                    CollectionOptions = new CreateCollectionOptions
                    {
                        CollectionName = "myCollection",
                        DatabaseOptions = new CreateDatabaseOptions
                        {
                            DatabaseName = "myDatabase"
                        }
                    },
                    CustomImportArgs = new []
                    {
                        "--type=tsv",
                        "--headerline"
                    }
                });

            // assert
            BsonDocument imported = collection.Find(new BsonDocument()).FirstOrDefault();
            imported.MatchSnapshot(o => o.IgnoreField("[0].Value"));
        }
    }
}
