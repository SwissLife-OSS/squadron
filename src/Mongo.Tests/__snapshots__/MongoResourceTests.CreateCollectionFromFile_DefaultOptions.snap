[
  {
    "Name": "_id",
    "Value": "5f4a348f673f5d8cd6fdc70d"
  },
  {
    "Name": "headline",
    "Value": "Apple Reported Fourth Quarter Revenue Today"
  },
  {
    "Name": "date",
    "Value": "2015-10-27T22:35:21.908Z"
  },
  {
    "Name": "views",
    "Value": 1132
  },
  {
    "Name": "author",
    "Value": [
      {
        "Name": "name",
        "Value": "Bob Walker"
      },
      {
        "Name": "title",
        "Value": "Lead Business Editor"
      }
    ]
  },
  {
    "Name": "published",
    "Value": true
  },
  {
    "Name": "tags",
    "Value": [
      "AAPL",
      [
        {
          "Name": "name",
          "Value": "city"
        },
        {
          "Name": "value",
          "Value": "Cupertino"
        }
      ],
      [
        "Electronics",
        "Computers"
      ]
    ]
  }
]
